code/go/0chain.net/smartcontract/partitions/partitions.go

```diff
@@ -514,26 +514,65 @@ func (p *Partitions) removeItem(
 	return p.loadLastFromPrev(state)
 }
 
-func (p *Partitions) GetRandomItems(state state.StateContextI, r *rand.Rand, vs interface{}) error {
+func (p *Partitions) GetRandomItems(balances state.StateContextI, r *rand.Rand, vs interface{}, prev, partitionSize int) (int, error) {
 	if p.Last.length() == 0 {
-		return errors.New("empty list, no items to return")
+		return -1, errors.New("empty list, no items to return")
 	}
 
-	var index int
+	var (
+		part      *partition
+		err       error
+		index     int
+		lastCount int
+	)
+
 	if p.Last.Loc > 0 {
-		index = r.Intn(p.Last.Loc + 1)
-	}
-	part, err := p.getPartition(state, index)
-	if err != nil {
-		return err
+		actErr := state.WithActivation(balances, "artemis", func() error {
+			index = r.Intn(p.Last.Loc + 1)
+			part, err = p.getPartition(balances, index)
+			if err != nil {
+				return err
+			}
+			return nil
+		}, func() error {
+			for {
+				index = r.Intn(p.Last.Loc + 1)
+
+				if index != prev {
```

Comment:
The new added `prev` params looks a bit weird for me.
We don't have to get rand index again and check if it's the index of previous one. Usually only the last partition could have number less than half of the partition size. So what we can do are:
1. check if it's the last partition, and if it is, then we check if it's less than half of the Size, if yes, then we get its previous partition if exists and get get the last `N` from the prev partition, `N = parititionsSize - items_num_in_last_part`. Then return the `append(prev_part_item, last_item)`. 
2. If the last partition is the only partition, then we just return all the items we have
3. If the selected part is not the last part, return the items in the selected part. 

With the changes above, we don't need to change the signature of `GetRandomItems()`, and the logic would be simpler.